File: src/modules/inventory/inventory.service.ts
import type { ClientSession } from 'mongoose';
import { AppError } from '../../shared/errors/AppError.js';
import { recordAudit } from '../../shared/utils/audit.js';
import {
  buildPaginationMeta,
  toPageOptions,
  type PaginationMeta,
} from '../../shared/utils/pagination.js';
import { Product } from '../catalog/product.model.js';
import { InventoryMovement, type MovementType } from './inventory.model.js';

export interface StockSnapshot {
  quantity: number;
  reserved: number;
  available: number;
  lowStockThreshold: number;
}

export interface MovementReference {
  referenceType: 'order' | 'orderItem' | 'manual';
  referenceId?: string;
}

interface MovementContext extends MovementReference {
  actorId?: string;
}

async function recordMovement(
  productId: string,
  variantId: string,
  sku: string,
  type: MovementType,
  quantity: number,
  beforeAvailable: number | null,
  afterAvailable: number | null,
  reason: string,
  context: MovementContext,
): Promise<void> {
  await InventoryMovement.create({
    productId,
    variantId,
    sku,
    type,
    quantity,
    beforeAvailable,
    afterAvailable,
    reason,
    referenceType: context.referenceType,
    referenceId: context.referenceId ?? null,
    actorId: context.actorId ?? null,
  });
}

function extractVariant(
  document: { variants: ReadonlyArray<{ _id?: unknown; sku: string; stock: StockSnapshot }> },
  variantId: string,
) {
  const variant = document.variants.find(
    (v) => (v._id as { toString(): string }).toString() === variantId,
  );
  return variant;
}

// ---------------------------------------------------------------------------
// Concurrency-safe stock operations
//
// Overselling is prevented at the database level: each operation is a single
// atomic findOneAndUpdate whose filter includes the stock guard (e.g.
// available >= qty). Two simultaneous reservations cannot both pass the guard,
// so it is impossible to reserve more than is available regardless of how many
// replicas/instances are running.
// ---------------------------------------------------------------------------

export async function reserveStock(
  productId: string,
  variantId: string,
  quantity: number,
  reason: string,
  context: MovementContext,
  session?: ClientSession,
): Promise<StockSnapshot> {
  const updated = await Product.findOneAndUpdate(
    {
      _id: productId,
      'variants._id': variantId,
      'variants.isActive': true,
      'variants.stock.available': { $gte: quantity },
    },
    { $inc: { 'variants.$.stock.reserved': quantity, 'variants.$.stock.available': -quantity } },
    { new: true, session },
  ).lean();

  if (!updated) {
    const exists = await Product.exists({ _id: productId, 'variants._id': variantId });
    if (!exists) throw AppError.notFound('Product variant not found');
    throw AppError.conflict('Insufficient stock for reservation', {
      errorCode: 'INSUFFICIENT_STOCK',
    });
  }

  const variant = extractVariant(updated, variantId);
  if (!variant) throw AppError.internal('Variant missing after stock update');
  const after = variant.stock.available;
  await recordMovement(
    productId,
    variantId,
    variant.sku,
    'reservation',
    -quantity,
    after + quantity,
    after,
    reason,
    context,
  );
  return variant.stock;
}

export async function releaseStock(
  productId: string,
  variantId: string,
  quantity: number,
  reason: string,
  context: MovementContext,
  session?: ClientSession,
): Promise<StockSnapshot> {
  const updated = await Product.findOneAndUpdate(
    {
      _id: productId,
      'variants._id': variantId,
      'variants.stock.reserved': { $gte: quantity },
    },
    { $inc: { 'variants.$.stock.reserved': -quantity, 'variants.$.stock.available': quantity } },
    { new: true, session },
  ).lean();

  if (!updated) {
    throw AppError.conflict('Cannot release more stock than is reserved', {
      errorCode: 'RESERVATION_MISMATCH',
    });
  }

  const variant = extractVariant(updated, variantId);
  if (!variant) throw AppError.internal('Variant missing after stock update');
  const after = variant.stock.available;
  await recordMovement(
    productId,
    variantId,
    variant.sku,
    'release',
    quantity,
    after - quantity,
    after,
    reason,
    context,
  );
  return variant.stock;
}

// A confirmed/paid order converts reservation into a sale: reserved decreases
// and total quantity decreases; available is untouched (it already dropped at
// reservation time).
export async function deductStock(
  productId: string,
  variantId: string,
  quantity: number,
  reason: string,
  context: MovementContext,
  session?: ClientSession,
): Promise<StockSnapshot> {
  const updated = await Product.findOneAndUpdate(
    {
      _id: productId,
      'variants._id': variantId,
      'variants.stock.reserved': { $gte: quantity },
    },
    { $inc: { 'variants.$.stock.reserved': -quantity, 'variants.$.stock.quantity': -quantity } },
    { new: true, session },
  ).lean();

  if (!updated) {
    throw AppError.conflict('Cannot deduct more stock than is reserved', {
      errorCode: 'RESERVATION_MISMATCH',
    });
  }

  const variant = extractVariant(updated, variantId);
  if (!variant) throw AppError.internal('Variant missing after stock update');
  await recordMovement(
    productId,
    variantId,
    variant.sku,
    'sale',
    -quantity,
    variant.stock.available,
    variant.stock.available,
    reason,
    context,
  );
  return variant.stock;
}

export async function restock(
  productId: string,
  variantId: string,
  quantity: number,
  reason: string,
  actorId: string,
): Promise<StockSnapshot> {
  const updated = await Product.findOneAndUpdate(
    { _id: productId, 'variants._id': variantId },
    { $inc: { 'variants.$.stock.quantity': quantity, 'variants.$.stock.available': quantity } },
    { new: true },
  ).lean();

  if (!updated) throw AppError.notFound('Product variant not found');

  const variant = extractVariant(updated, variantId);
  if (!variant) throw AppError.internal('Variant missing after stock update');
  const after = variant.stock.available;
  await recordMovement(
    productId,
    variantId,
    variant.sku,
    'restock',
    quantity,
    after - quantity,
    after,
    reason,
    { referenceType: 'manual', actorId },
  );
  recordAudit(
    'inventory.restocked',
    { productId, variantId, quantity, sku: variant.sku },
    { actorId },
  );
  return variant.stock;
}

// Manual reconciliation: sets the physical quantity to an absolute value and
// adjusts available by the same delta. Deliberately not guarded atomically —
// manual corrections are low-concurrency admin actions, unlike reservations.
export async function adjustStock(
  productId: string,
  variantId: string,
  newQuantity: number,
  reason: string,
  actorId: string,
): Promise<StockSnapshot> {
  const product = await Product.findOne({ _id: productId, 'variants._id': variantId });
  if (!product) throw AppError.notFound('Product variant not found');

  const current = extractVariant(product, variantId);
  if (!current) throw AppError.internal('Variant missing');

  const delta = newQuantity - current.stock.quantity;
  if (delta === 0) return current.stock;

  const newAvailable = current.stock.available + delta;
  if (newAvailable < 0) {
    throw AppError.badRequest(
      'Adjustment would make available stock negative; release reservations first',
      { errorCode: 'NEGATIVE_AVAILABLE_STOCK' },
    );
  }

  const updated = await Product.findOneAndUpdate(
    { _id: productId, 'variants._id': variantId },
    {
      $inc: { 'variants.$.stock.quantity': delta, 'variants.$.stock.available': delta },
    },
    { new: true },
  ).lean();
  if (!updated) throw AppError.internal('Stock update failed');

  const variant = extractVariant(updated, variantId);
  if (!variant) throw AppError.internal('Variant missing after stock update');
  const after = variant.stock.available;
  await recordMovement(
    productId,
    variantId,
    variant.sku,
    'adjustment',
    delta,
    after - delta,
    after,
    reason,
    { referenceType: 'manual', actorId },
  );
  recordAudit('inventory.adjusted', { productId, variantId, delta, reason }, { actorId });
  return variant.stock;
}

export async function getStock(productId: string, variantId: string): Promise<StockSnapshot> {
  const product = await Product.findOne({ _id: productId, 'variants._id': variantId }).lean();
  if (!product) throw AppError.notFound('Product variant not found');
  const variant = extractVariant(product, variantId);
  if (!variant) throw AppError.notFound('Product variant not found');
  return variant.stock;
}

export interface MovementPublic {
  id: string;
  productId: string;
  variantId: string;
  sku: string;
  type: MovementType;
  quantity: number;
  beforeAvailable: number | null;
  afterAvailable: number | null;
  reason: string;
  referenceType: string | null;
  referenceId: string | null;
  createdAt: string;
}

export async function listMovements(query: {
  page?: number;
  limit?: number;
  productId?: string;
  variantId?: string;
  type?: MovementType;
  sku?: string;
}): Promise<{ movements: MovementPublic[]; meta: PaginationMeta }> {
  const filter: Record<string, unknown> = {};
  if (query.productId) filter.productId = query.productId;
  if (query.variantId) filter.variantId = query.variantId;
  if (query.type) filter.type = query.type;
  if (query.sku) filter.sku = query.sku.toUpperCase();

  const options = toPageOptions(query);
  const [total, movements] = await Promise.all([
    InventoryMovement.countDocuments(filter),
    InventoryMovement.find(filter)
      .sort({ createdAt: -1 })
      .skip(options.skip)
      .limit(options.limit)
      .lean(),
  ]);

  return {
    movements: movements.map((m) => ({
      id: m._id.toString(),
      productId: m.productId.toString(),
      variantId: m.variantId.toString(),
      sku: m.sku,
      type: m.type,
      quantity: m.quantity,
      beforeAvailable: m.beforeAvailable,
      afterAvailable: m.afterAvailable,
      reason: m.reason,
      referenceType: m.referenceType,
      referenceId: m.referenceId,
      createdAt: m.createdAt.toISOString(),
    })),
    meta: buildPaginationMeta(total, options),
  };
}

export interface LowStockItem {
  productId: string;
  name: string;
  sku: string;
  variantId: string;
  available: number;
  lowStockThreshold: number;
}

// Any active variant whose available stock has dropped to or below its
// threshold. Used by the admin dashboard and the low-stock notification job.
export async function listLowStock(): Promise<LowStockItem[]> {
  const products = await Product.aggregate([
    {
      $match: {
        status: 'published',
        isActive: true,
        variants: { $elemMatch: { isActive: true } },
      },
    },
    { $unwind: '$variants' },
    {
      $match: {
        'variants.isActive': true,
        $expr: { $lte: ['$variants.stock.available', '$variants.stock.lowStockThreshold'] },
      },
    },
    { $sort: { 'variants.stock.available': 1 } },
  ]);

  return products.map((p) => ({
    productId: p._id.toString(),
    name: p.name,
    sku: p.variants.sku,
    variantId: p.variants._id.toString(),
    available: p.variants.stock.available,
    lowStockThreshold: p.variants.stock.lowStockThreshold,
  }));
}

export type { MovementType };

File: java/src/main/java/com/ecommerce/backend/modules/inventory/InventoryService.java
package com.ecommerce.backend.modules.inventory;

import com.ecommerce.backend.common.error.ApiException;
import com.ecommerce.backend.modules.audit.AuditService;
import com.ecommerce.backend.modules.catalog.Product;
import com.ecommerce.backend.modules.catalog.Variant;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.bson.Document;
import org.springframework.data.domain.Sort;
import org.springframework.data.mongodb.core.FindAndModifyOptions;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.aggregation.Aggregation;
import org.springframework.data.mongodb.core.aggregation.AggregationOperation;
import org.springframework.data.mongodb.core.aggregation.AggregationResults;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.data.mongodb.core.query.Update;
import org.springframework.stereotype.Service;

/**
 * Concurrency-safe stock operations (mirrors {@code inventory.service.ts}).
 * Each operation is a single atomic findAndModify whose filter includes the
 * stock guard (e.g. available &gt;= qty), so overselling is prevented at the
 * database level even across replicas.
 */
@Service
public class InventoryService {

    private final MongoTemplate mongo;
    private final InventoryMovementRepository movementRepository;
    private final AuditService auditService;

    public InventoryService(
            MongoTemplate mongo,
            InventoryMovementRepository movementRepository,
            AuditService auditService) {
        this.mongo = mongo;
        this.movementRepository = movementRepository;
        this.auditService = auditService;
    }

    public record StockSnapshot(
            long quantity, long reserved, long available, long lowStockThreshold) {}

    public record MovementReference(String referenceType, String referenceId, String actorId) {
        public static MovementReference manual(String actorId) {
            return new MovementReference("manual", null, actorId);
        }
    }

    // ------------------------------------------------------------------ Ops

    public StockSnapshot reserveStock(
            String productId, String variantId, long quantity, String reason, MovementReference ctx) {
        Product updated = findAndModify(
                new Query(Criteria.where("_id")
                        .is(productId)
                        .and("variants._id")
                        .is(variantId)
                        .and("variants.isActive")
                        .is(true)
                        .and("variants.stock.available")
                        .gte(quantity)),
                new Update()
                        .inc("variants.$.stock.reserved", quantity)
                        .inc("variants.$.stock.available", -quantity));

        if (updated == null) {
            boolean exists = variantExists(productId, variantId);
            if (!exists) {
                throw ApiException.notFound("Product variant not found");
            }
            throw ApiException.conflict("Insufficient stock for reservation", "INSUFFICIENT_STOCK");
        }
        Variant variant = requireVariant(updated, variantId);
        long after = variant.available();
        recordMovement(productId, variantId, variant.sku, InventoryMovement.TYPE_RESERVATION,
                -quantity, after + quantity, after, reason, ctx);
        return snapshot(variant);
    }

    public StockSnapshot releaseStock(
            String productId, String variantId, long quantity, String reason, MovementReference ctx) {
        Product updated = findAndModify(
                new Query(Criteria.where("_id")
                        .is(productId)
                        .and("variants._id")
                        .is(variantId)
                        .and("variants.stock.reserved")
                        .gte(quantity)),
                new Update()
                        .inc("variants.$.stock.reserved", -quantity)
                        .inc("variants.$.stock.available", quantity));

        if (updated == null) {
            throw ApiException.conflict(
                    "Cannot release more stock than is reserved", "RESERVATION_MISMATCH");
        }
        Variant variant = requireVariant(updated, variantId);
        long after = variant.available();
        recordMovement(productId, variantId, variant.sku, InventoryMovement.TYPE_RELEASE,
                quantity, after - quantity, after, reason, ctx);
        return snapshot(variant);
    }

    /** A confirmed/paid order converts reservation into a sale. */
    public StockSnapshot deductStock(
            String productId, String variantId, long quantity, String reason, MovementReference ctx) {
        Product updated = findAndModify(
                new Query(Criteria.where("_id")
                        .is(productId)
                        .and("variants._id")
                        .is(variantId)
                        .and("variants.stock.reserved")
                        .gte(quantity)),
                new Update()
                        .inc("variants.$.stock.reserved", -quantity)
                        .inc("variants.$.stock.quantity", -quantity));

        if (updated == null) {
            throw ApiException.conflict(
                    "Cannot deduct more stock than is reserved", "RESERVATION_MISMATCH");
        }
        Variant variant = requireVariant(updated, variantId);
        recordMovement(productId, variantId, variant.sku, InventoryMovement.TYPE_SALE,
                -quantity, variant.available(), variant.available(), reason, ctx);
        return snapshot(variant);
    }

    public StockSnapshot restock(
            String productId, String variantId, long quantity, String reason, String actorId) {
        Product updated = findAndModify(
                new Query(Criteria.where("_id").is(productId).and("variants._id").is(variantId)),
                new Update()
                        .inc("variants.$.stock.quantity", quantity)
                        .inc("variants.$.stock.available", quantity));

        if (updated == null) {
            throw ApiException.notFound("Product variant not found");
        }
        Variant variant = requireVariant(updated, variantId);
        long after = variant.available();
        recordMovement(productId, variantId, variant.sku, InventoryMovement.TYPE_RESTOCK,
                quantity, after - quantity, after, reason, MovementReference.manual(actorId));
        auditService.log("inventory.restocked", actorId, null,
                Map.of("productId", productId, "variantId", variantId,
                        "quantity", quantity, "sku", variant.sku));
        return snapshot(variant);
    }

    public StockSnapshot adjustStock(
            String productId, String variantId, long newQuantity, String reason, String actorId) {
        Product product = mongo.findOne(
                Query.query(Criteria.where("_id").is(productId).and("variants._id").is(variantId)),
                Product.class);
        if (product == null) {
            throw ApiException.notFound("Product variant not found");
        }
        Variant current = requireVariant(product, variantId);

        long delta = newQuantity - current.stock.quantity;
        if (delta == 0) {
            return snapshot(current);
        }
        long newAvailable = current.available() + delta;
        if (newAvailable < 0) {
            throw ApiException.badRequest(
                    "Adjustment would make available stock negative; release reservations first",
                    "NEGATIVE_AVAILABLE_STOCK");
        }

        Product updated = findAndModify(
                new Query(Criteria.where("_id").is(productId).and("variants._id").is(variantId)),
                new Update()
                        .inc("variants.$.stock.quantity", delta)
                        .inc("variants.$.stock.available", delta));
        if (updated == null) {
            throw ApiException.badRequest("Stock update failed", "STOCK_UPDATE_FAILED");
        }
        Variant variant = requireVariant(updated, variantId);
        long after = variant.available();
        recordMovement(productId, variantId, variant.sku, InventoryMovement.TYPE_ADJUSTMENT,
                delta, after - delta, after, reason, MovementReference.manual(actorId));
        auditService.log("inventory.adjusted", actorId, null,
                Map.of("productId", productId, "variantId", variantId, "delta", delta, "reason", reason));
        return snapshot(variant);
    }

    public StockSnapshot getStock(String productId, String variantId) {
        Product product = mongo.findOne(
                Query.query(Criteria.where("_id").is(productId).and("variants._id").is(variantId)),
                Product.class);
        if (product == null) {
            throw ApiException.notFound("Product variant not found");
        }
        return snapshot(requireVariant(product, variantId));
    }

    // -------------------------------------------------------------- Movements

    public record MovementPublic(
            String id,
            String productId,
            String variantId,
            String sku,
            String type,
            long quantity,
            Long beforeAvailable,
            Long afterAvailable,
            String reason,
            String referenceType,
            String referenceId,
            String createdAt) {}

    public record MovementListResult(List<MovementPublic> movements, long total) {}

    public MovementListResult listMovements(
            int page,
            int limit,
            String productId,
            String variantId,
            String type,
            String sku) {
        Query query = new Query();
        if (productId != null && !productId.isBlank()) {
            query.addCriteria(Criteria.where("productId").is(productId));
        }
        if (variantId != null && !variantId.isBlank()) {
            query.addCriteria(Criteria.where("variantId").is(variantId));
        }
        if (type != null && !type.isBlank()) {
            query.addCriteria(Criteria.where("type").is(type));
        }
        if (sku != null && !sku.isBlank()) {
            query.addCriteria(Criteria.where("sku").is(sku.toUpperCase()));
        }
        long total = mongo.count(query, InventoryMovement.class);
        query.with(Sort.by(Sort.Direction.DESC, "createdAt"))
                .skip((long) (page - 1) * limit)
                .limit(limit);
        List<InventoryMovement> movements = mongo.find(query, InventoryMovement.class);
        List<MovementPublic> items = movements.stream()
                .map(m -> new MovementPublic(
                        m.id,
                        m.productId,
                        m.variantId,
                        m.sku,
                        m.type,
                        m.quantity,
                        m.beforeAvailable,
                        m.afterAvailable,
                        m.reason,
                        m.referenceType,
                        m.referenceId,
                        iso(m.createdAt)))
                .toList();
        return new MovementListResult(items, total);
    }

    public record LowStockItem(
            String productId, String name, String sku, String variantId,
            long available, long lowStockThreshold) {}

    public List<LowStockItem> listLowStock() {
        List<AggregationOperation> ops = new ArrayList<>();
        ops.add(ctx -> new Document("$match", new Document()
                .append("status", Product.STATUS_PUBLISHED)
                .append("isActive", true)
                .append("variants", new Document("$elemMatch", new Document("isActive", true)))));
        ops.add(ctx -> new Document("$unwind", "$variants"));
        ops.add(ctx -> new Document("$match", new Document()
                .append("variants.isActive", true)
                .append("$expr", new Document("$lte", List.of(
                        "$variants.stock.available",
                        "$variants.stock.lowStockThreshold")))));
        ops.add(ctx -> new Document("$sort", new Document("variants.stock.available", 1)));

        AggregationResults<Document> result = mongo.aggregate(
                Aggregation.newAggregation(Product.class, ops), Document.class);
        return result.getMappedResults().stream()
                .map(d -> {
                    Document variant = (Document) d.get("variants");
                    Document stock = (Document) variant.get("stock");
                    return new LowStockItem(
                            d.get("_id").toString(),
                            d.getString("name"),
                            variant.getString("sku"),
                            variant.get("_id").toString(),
                            stock.get("available", Number.class).longValue(),
                            stock.get("lowStockThreshold", Number.class).longValue());
                })
                .toList();
    }

    // -------------------------------------------------------------- Internals

    private Product findAndModify(Query query, Update update) {
        return mongo.findAndModify(
                query, update, FindAndModifyOptions.options().returnNew(true), Product.class);
    }

    private boolean variantExists(String productId, String variantId) {
        return mongo.exists(
                Query.query(Criteria.where("_id").is(productId).and("variants._id").is(variantId)),
                Product.class);
    }

    private Variant requireVariant(Product product, String variantId) {
        for (Variant variant : product.variants) {
            if (variantId.equals(variant.id)) {
                return variant;
            }
        }
        throw ApiException.badRequest("Variant missing after stock update", "VARIANT_MISSING");
    }

    private StockSnapshot snapshot(Variant variant) {
        return new StockSnapshot(
                variant.stock.quantity,
                variant.stock.reserved,
                variant.stock.available,
                variant.stock.lowStockThreshold);
    }

    private void recordMovement(
            String productId,
            String variantId,
            String sku,
            String type,
            long quantity,
            Long before,
            Long after,
            String reason,
            MovementReference ctx) {
        try {
            InventoryMovement movement = new InventoryMovement();
            movement.productId = productId;
            movement.variantId = variantId;
            movement.sku = sku;
            movement.type = type;
            movement.quantity = quantity;
            movement.beforeAvailable = before;
            movement.afterAvailable = after;
            movement.reason = reason;
            movement.referenceType = ctx.referenceType();
            movement.referenceId = ctx.referenceId();
            movement.actorId = ctx.actorId();
            movement.createdAt = Instant.now();
            movementRepository.insert(movement);
        } catch (RuntimeException e) {
            // Stock bookkeeping must never break the operation that triggered it.
        }
    }

    private static String iso(Instant value) {
        return value == null ? null : value.toString();
    }
}
